File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/playground/KnobControls.kt
package xyz.ksharma.prisma.catalogue.playground

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.components.switchctl.PrismaSwitch
import xyz.ksharma.prisma.components.textfield.PrismaTextField
import xyz.ksharma.prisma.coreui.themed
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

@Composable
internal fun KnobLabel(text: String, hint: String? = null) {
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        Text(
            text = text,
            style = PrismaTypography.LabelMd,
            color = PrismaSemanticColors.TextSecondary.themed(),
        )
        if (hint != null) {
            Text(
                text = hint,
                style = PrismaTypography.BodySm,
                color = PrismaSemanticColors.TextTertiary.themed(),
            )
        }
    }
}

/** A single editable text knob — bound to a [PrismaTextField]. */
@Composable
public fun StringKnobRow(
    label: String,
    value: String,
    onChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "",
    helper: String? = null,
) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2)) {
        KnobLabel(label, helper)
        PrismaTextField(
            value = value,
            onValueChange = onChange,
            placeholder = placeholder.takeIf { it.isNotEmpty() },
        )
    }
}

/** Boolean knob — label on left, [PrismaSwitch] on right. */
@Composable
public fun BoolKnobRow(
    label: String,
    value: Boolean,
    onChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
    helper: String? = null,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp4),
    ) {
        Box(modifier = Modifier.weight(1f)) { KnobLabel(label, helper) }
        PrismaSwitch(checked = value, onCheckedChange = onChange)
    }
}

/** Numeric knob backed by a slider. */
@Composable
public fun IntKnobRow(
    label: String,
    value: Int,
    range: IntRange,
    onChange: (Int) -> Unit,
    modifier: Modifier = Modifier,
    step: Int = 1,
) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            KnobLabel(label)
            Text(
                text = value.toString(),
                style = PrismaTypography.LabelMd,
                color = PrismaSemanticColors.TextPrimary.themed(),
            )
        }
        val steps = if (step > 0) ((range.last - range.first) / step - 1).coerceAtLeast(0) else 0
        Slider(
            value = value.toFloat(),
            onValueChange = { onChange(it.toInt()) },
            valueRange = range.first.toFloat()..range.last.toFloat(),
            steps = steps,
        )
    }
}

/**
 * Enum knob rendered as a segmented chip row. For long enum lists this still
 * scrolls horizontally via [FlowRow] wrapping to multi-line if needed.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
public fun <T> EnumKnobRow(
    label: String,
    value: T,
    values: List<T>,
    onChange: (T) -> Unit,
    modifier: Modifier = Modifier,
    optionLabel: (T) -> String = { it.toString() },
) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2)) {
        KnobLabel(label)
        FlowRow(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
            verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
        ) {
            values.forEach { v ->
                EnumChip(
                    label = optionLabel(v),
                    selected = v == value,
                    onClick = { onChange(v) },
                )
            }
        }
    }
}

@Composable
private fun EnumChip(label: String, selected: Boolean, onClick: () -> Unit) {
    val bg = if (selected) PrismaSemanticColors.AccentDefault.themed()
             else PrismaSemanticColors.SurfaceRaised.themed()
    val fg = if (selected) PrismaSemanticColors.TextOnAccent.themed()
             else PrismaSemanticColors.TextPrimary.themed()
    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(PrismaRadius.Full))
            .background(bg)
            .let {
                if (!selected) it.border(
                    1.dp,
                    PrismaSemanticColors.BorderSubtle.themed(),
                    RoundedCornerShape(PrismaRadius.Full),
                ) else it
            }
            .clickable(role = Role.Button, onClick = onClick)
            .padding(horizontal = PrismaSpacing.Sp3, vertical = PrismaSpacing.Sp2),
    ) {
        Text(text = label, style = PrismaTypography.LabelSm, color = fg)
    }
}

/**
 * Icon picker knob — renders selectable icon tiles. The first tile is "None"
 * (returns `null`) when [nullable] is true.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
public fun IconKnobRow(
    label: String,
    value: Int?,
    options: List<Pair<String, Int>>,
    onChange: (Int?) -> Unit,
    modifier: Modifier = Modifier,
    nullable: Boolean = true,
) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2)) {
        KnobLabel(label)
        FlowRow(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
            verticalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
        ) {
            if (nullable) {
                IconTile(
                    selected = value == null,
                    onClick = { onChange(null) },
                    iconRes = null,
                    fallback = "—",
                )
            }
            options.forEach { (name, res) ->
                IconTile(
                    selected = value == res,
                    onClick = { onChange(res) },
                    iconRes = res,
                    fallback = name,
                )
            }
        }
    }
}

@Composable
private fun IconTile(
    selected: Boolean,
    onClick: () -> Unit,
    iconRes: Int?,
    fallback: String,
) {
    val bg = if (selected) PrismaSemanticColors.AccentSubtle.themed()
             else PrismaSemanticColors.SurfaceRaised.themed()
    val borderColor = if (selected) PrismaSemanticColors.AccentDefault.themed()
                     else PrismaSemanticColors.BorderSubtle.themed()
    Box(
        modifier = Modifier
            .size(40.dp)
            .clip(RoundedCornerShape(PrismaRadius.Md))
            .background(bg)
            .border(
                if (selected) 2.dp else 1.dp,
                borderColor,
                RoundedCornerShape(PrismaRadius.Md),
            )
            .clickable(role = Role.Button, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        if (iconRes != null) {
            Icon(
                painter = painterResource(iconRes),
                contentDescription = fallback,
                tint = PrismaSemanticColors.TextPrimary.themed(),
                modifier = Modifier.size(18.dp),
            )
        } else {
            Text(
                text = fallback,
                style = PrismaTypography.LabelMd,
                color = PrismaSemanticColors.TextTertiary.themed(),
            )
        }
    }
}

File: ios-catalogue/App/Playground/KnobControls.swift
import SwiftUI
import CoreUI
import Components

private struct KnobLabel: View {
    @Environment(\.colorScheme) private var scheme
    let text: String
    var hint: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(text)
                .font(PrismaTypography.labelMd.font)
                .foregroundStyle(PrismaSemanticColors.textSecondary.themed(scheme))
            if let hint {
                Text(hint)
                    .font(PrismaTypography.bodySm.font)
                    .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
            }
        }
    }
}

/// Editable text knob, bound to a `PrismaTextField`.
struct StringKnobRow: View {
    let label: String
    @Binding var value: String
    var placeholder: String = ""
    var helper: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp2) {
            KnobLabel(text: label, hint: helper)
            PrismaTextField(text: $value, placeholder: placeholder.isEmpty ? nil : placeholder)
        }
    }
}

/// Boolean knob — label on the left, toggle on the right.
struct BoolKnobRow: View {
    @Environment(\.colorScheme) private var scheme
    let label: String
    @Binding var value: Bool
    var helper: String? = nil

    var body: some View {
        HStack(spacing: PrismaSpacing.sp4) {
            KnobLabel(text: label, hint: helper)
            Spacer()
            Toggle("", isOn: $value)
                .labelsHidden()
                .tint(PrismaSemanticColors.accentDefault.themed(scheme))
        }
    }
}

/// Numeric knob backed by a slider.
struct IntKnobRow: View {
    @Environment(\.colorScheme) private var scheme
    let label: String
    @Binding var value: Int
    let range: ClosedRange<Int>
    var step: Int = 1

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp2) {
            HStack {
                KnobLabel(text: label)
                Spacer()
                Text("\(value)")
                    .font(PrismaTypography.labelMd.font)
                    .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
            }
            Slider(
                value: Binding(
                    get: { Double(value) },
                    set: { value = Int($0) }
                ),
                in: Double(range.lowerBound)...Double(range.upperBound),
                step: Double(step)
            )
            .tint(PrismaSemanticColors.accentDefault.themed(scheme))
        }
    }
}

/// Enum knob rendered as a chip row.
struct EnumKnobRow<Value: Hashable>: View {
    let label: String
    @Binding var value: Value
    let values: [Value]
    var optionLabel: (Value) -> String = { String(describing: $0) }

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp2) {
            KnobLabel(text: label)
            FlowLayout(spacing: PrismaSpacing.sp2) {
                ForEach(values, id: \.self) { v in
                    EnumChip(
                        label: optionLabel(v),
                        selected: v == value,
                        action: { value = v }
                    )
                }
            }
        }
    }
}

private struct EnumChip: View {
    @Environment(\.colorScheme) private var scheme
    let label: String
    let selected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(label)
                .font(PrismaTypography.labelSm.font)
                .padding(.horizontal, PrismaSpacing.sp3)
                .padding(.vertical, PrismaSpacing.sp2)
                .foregroundStyle(
                    selected
                        ? PrismaSemanticColors.textOnAccent.themed(scheme)
                        : PrismaSemanticColors.textPrimary.themed(scheme)
                )
                .background(
                    selected
                        ? PrismaSemanticColors.accentDefault.themed(scheme)
                        : PrismaSemanticColors.surfaceRaised.themed(scheme)
                )
                .clipShape(Capsule())
                .overlay(
                    Capsule()
                        .strokeBorder(
                            selected ? Color.clear : PrismaSemanticColors.borderSubtle.themed(scheme),
                            lineWidth: 1
                        )
                )
        }
        .buttonStyle(.plain)
    }
}

/// Icon picker knob. The first tile is "None" (returns nil) when [nullable] is true.
struct IconKnobRow: View {
    let label: String
    @Binding var value: PrismaIcon?
    let options: [PrismaIcon]
    var nullable: Bool = true

    var body: some View {
        VStack(alignment: .leading, spacing: PrismaSpacing.sp2) {
            KnobLabel(text: label)
            FlowLayout(spacing: PrismaSpacing.sp2) {
                if nullable {
                    IconTile(selected: value == nil, action: { value = nil }, icon: nil, fallback: "—")
                }
                ForEach(options, id: \.self) { icon in
                    IconTile(selected: value == icon, action: { value = icon }, icon: icon, fallback: icon.rawValue)
                }
            }
        }
    }
}

private struct IconTile: View {
    @Environment(\.colorScheme) private var scheme
    let selected: Bool
    let action: () -> Void
    let icon: PrismaIcon?
    let fallback: String

    var body: some View {
        Button(action: action) {
            Group {
                if let icon {
                    Image(prisma: icon)
                        .renderingMode(.template)
                        .resizable()
                        .frame(width: 18, height: 18)
                        .foregroundStyle(PrismaSemanticColors.textPrimary.themed(scheme))
                } else {
                    Text(fallback)
                        .font(PrismaTypography.labelMd.font)
                        .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
                }
            }
            .frame(width: 40, height: 40)
            .background(
                selected
                    ? PrismaSemanticColors.accentSubtle.themed(scheme)
                    : PrismaSemanticColors.surfaceRaised.themed(scheme)
            )
            .clipShape(RoundedRectangle(cornerRadius: PrismaRadius.md))
            .overlay(
                RoundedRectangle(cornerRadius: PrismaRadius.md)
                    .strokeBorder(
                        selected
                            ? PrismaSemanticColors.accentDefault.themed(scheme)
                            : PrismaSemanticColors.borderSubtle.themed(scheme),
                        lineWidth: selected ? 2 : 1
                    )
            )
        }
        .buttonStyle(.plain)
    }
}

// MARK: - FlowLayout

/// Minimal flow layout (SwiftUI's stock HStack doesn't wrap). Used by knob
/// controls that produce a variable number of chip-like children.
struct FlowLayout: Layout {
    let spacing: CGFloat

    init(spacing: CGFloat = 8) {
        self.spacing = spacing
    }

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var lineWidth: CGFloat = 0
        var lineHeight: CGFloat = 0
        var totalHeight: CGFloat = 0
        var totalWidth: CGFloat = 0
        for subview in subviews {
            let s = subview.sizeThatFits(.unspecified)
            if lineWidth + s.width > maxWidth {
                totalHeight += lineHeight + spacing
                totalWidth = max(totalWidth, lineWidth - spacing)
                lineWidth = s.width + spacing
                lineHeight = s.height
            } else {
                lineWidth += s.width + spacing
                lineHeight = max(lineHeight, s.height)
            }
        }
        totalHeight += lineHeight
        totalWidth = max(totalWidth, lineWidth - spacing)
        return CGSize(width: totalWidth, height: totalHeight)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX
        var y = bounds.minY
        var lineHeight: CGFloat = 0
        for subview in subviews {
            let s = subview.sizeThatFits(.unspecified)
            if x + s.width > bounds.maxX {
                x = bounds.minX
                y += lineHeight + spacing
                lineHeight = 0
            }
            subview.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(s))
            x += s.width + spacing
            lineHeight = max(lineHeight, s.height)
        }
    }
}
